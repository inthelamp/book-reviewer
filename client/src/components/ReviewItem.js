import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types'
import { Button } from 'react-bootstrap'

/**
 * @typedef {Object} Review
 * @property {string} id - Review id
 * @property {string} title - Review title
 * @property {string} status - Review status
 */ 
/**
 * Displaying review title and button
 * @param {Review} review - Review object 
 */
const ReviewItem = ( { review } ) => {
    const history = useHistory();

    const onSubmit = (e) => {        
        e.preventDefault()

        history.push(
            { 
                pathname: '/',
                state: { Review: review }
            }
        )
    }

    return ( 
        <div>
            {review.title} &nbsp;
            <Button className="float-end" variant='secondary' size="sm" onClick={(e) => onSubmit(e)}>
                Read
            </Button>
        </div>
    )
}

ReviewItem.propTypes = {
    review: PropTypes.shape({
            id: PropTypes.string,
            title: PropTypes.string,
            status: PropTypes.string        
    })
}

export default  ReviewItem