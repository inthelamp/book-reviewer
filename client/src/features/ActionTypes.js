const ActionTypes = {
	SIGNED_IN: 'signedIn',
	SIGNED_OUT: 'signedOut'
}

export default ActionTypes