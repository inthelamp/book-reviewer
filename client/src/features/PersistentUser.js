import jwt_decode from 'jwt-decode'

export const getToken = () => {
    const name = 'token='
    const decodedCookie = decodeURIComponent(document.cookie)
    const cookieAttributes = decodedCookie.split(';')

    for(let i = 0; i <cookieAttributes.length; i++) {
        let ca = cookieAttributes[i]
        while (ca.charAt(0) === ' ') {
            ca = ca.substring(1)
        }
        if (ca.indexOf(name) === 0) {
            return ca.substring(name.length, ca.length)
        }
    }
    return ''
}

export const isTokenExpried = () => {
    const token = getToken()

    if (token) {      
        try {
            const decoded = jwt_decode(token)     
            const { exp } = decoded

            if (Date.now() < exp * 1000) { 
                return false
            } 
        } catch (error) {
            console.log(error)
        }
    }   

    return true
}

export const persistentUser = () => {
    const token = getToken()

    if (token) {      
        try {
            const decoded = jwt_decode(token)     
            const { name, role, exp } = decoded

            if (Date.now() < exp * 1000) { 
                return {
                    name,
                    role,
                    "isSignedIn": true
                }
            } else {
                return {
                    "isTokenExpried": true
                }
            }
        } catch (error) {
            console.log(error)
        }
    }

    return null
}