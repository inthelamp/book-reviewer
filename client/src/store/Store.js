import { createStore, combineReducers } from 'redux'
import UserReducer from './UserReducer'


//const rootReducer = combineReducers({User: UserReducer})
const Store = createStore(UserReducer)

export default Store