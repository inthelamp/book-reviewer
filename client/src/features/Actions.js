import { User, TextEditor, Review } from './ActionTypes'

/**
 * SignedIn action
 * @param {string} id - user id
 * @param {string} name - user name
 * @param {string} role - user role
 */
export const SignedIn = (id, name, role) => ({
    type: User.SIGNED_IN,
    payload: {
        id,
        name, 
        role
    }
})

// SignedOut action
export const SignedOut = () => ({
    type: User.SIGNED_OUT
})

// ChangedModeToBrowse action
export const ChangedModeToBrowse = (review) => ({
    type: TextEditor.CHANGED_MODE_TO_BROWSE,
    payload: {
        review
    }
})

// ChangedModeToEdit action
export const ChangedModeToEdit = (review) => ({
    type: TextEditor.CHANGED_MODE_TO_EDIT,
    payload: {
        review
    }
})

// PostedReview action
export const PostedReview = () => ({
    type: Review.POSTED_REVIEW
})

// ListedReview action
export const ListedReview = () => ({
    type: Review.LISTED_REVIEW
})