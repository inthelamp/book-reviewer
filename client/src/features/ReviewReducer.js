import { Review } from './ActionTypes'

// Review reducer
const ReviewReducer = (state = [], action) => {
    switch(action.type) {
        case Review.POSTED_REVIEW:
            return {
                ...state, 
                status: Review.POSTED_REVIEW,         
            }
            
        case Review.LISTED_REVIEW:
            return {
                ...state, 
                status: Review.LISTED_REVIEW,         
            }
            
        default:
            return state    
    }
}

export default ReviewReducer
