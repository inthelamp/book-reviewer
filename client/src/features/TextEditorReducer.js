import { TextEditor } from './ActionTypes'

// TextEditor reducer
const TextEditorReducer = (state = [], action) => {
    switch(action.type) {
        case TextEditor.CHANGED_MODE_TO_BROWSE:
            return {
                ...state, 
                review: action.payload.review,   
                mode: TextEditor.CHANGED_MODE_TO_BROWSE,           
            }
            
        case TextEditor.CHANGED_MODE_TO_EDIT:
            return {
                ...state, 
                review: action.payload.review,                
                mode: TextEditor.CHANGED_MODE_TO_EDIT
            }
            
        default:
            return state    
    }
}

export default TextEditorReducer
