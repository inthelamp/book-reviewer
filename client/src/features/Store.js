import { createStore, combineReducers } from 'redux'
import UserReducer from './UserReducer'
import TextEditorReducer from './TextEditorReducer'

// Creating store
const rootReducer = combineReducers({ User: UserReducer, TextEditor: TextEditorReducer})
const Store = createStore(
        rootReducer,
        window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() 
)

export default Store