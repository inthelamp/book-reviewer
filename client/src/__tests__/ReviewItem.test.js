import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import Store from '../features/Store'
import ReviewItem from '../components/ReviewItem'
import { Review } from '../features/ActionTypes'

describe('Component: ReviewItem', () => {
    it('Component: ReviewItem rendering correctly', () => {  
        const review = {
            reviewId: 'review-id-xxxxxx',
            subject: 'test subject',
            isOwner: true,
            status: Review.POSTED_REVIEW
        }

        const tree = renderer
          .create(<Provider store={Store}><ReviewItem review={review} /></Provider>)
          .toJSON();
        expect(tree).toMatchSnapshot()
    })
})