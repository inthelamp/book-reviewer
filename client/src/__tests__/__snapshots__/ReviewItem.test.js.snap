// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ReviewItem Component: ReviewItem rendering correctly 1`] = `
<div>
  test subject
   Â 
  <button
    className="float-end btn btn-secondary btn-sm"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Read
  </button>
</div>
`;
